mapping: alloc_mapping_huge_page_aligned for SHM

bartoldeman opened the pull request:
This change introduces alloc_mapping_huge_page_aligned and removes the now redundant alias_mapping_huge_page_aligned. It's used for lowmem_base and vga.mem.base, and hence the intermediate mapping for the VGAEMU LFB is no longer needed for KVM.

The guard page logic that was used for vga.mem.base in vgaemu.c has been moved into the mapping functions, so that there's always a PROT_NOTE page in front of a huge page aligned mmap and SHM alloc, and one after a huge page aligned SHM alloc as well.

See also #1922

bartoldeman committed:
mapping: directly map shared memory

Passing target = (void *)-1 we can skip the intermediate private
anonymous mapping.

bartoldeman commented:
This is against devel for now but I can rebase later when needed. It's getting late here, will have another look at #1922 tomorrow.

stsp commented on src/arch/linux/mapping/mapfile.c:
```diff
@@ -124,8 +125,10 @@ static int open_mapping_f(int cap)
     */
-    mpool = mmap(0, mapsize, PROT_READ|PROT_WRITE,
-    		MAP_SHARED, tmpfile_fd, 0);
+    mpool = mmap_mapping_huge_page_aligned(cap, mapsize, PROT_NONE);
```
I wonder why its a good idea.
Putting shm in aligned block doesn't
mean every other alloc will be similar
aligned.

bartoldeman replied:
True, I just did that to waste a bit less space. This way the lowmem_base alloc can go right at the beginning, then a little hole, then the vgaemu map, and the rest (EMS/XMS) is allocated on demand between the first two or after vgaemu. But it may not matter all that much in the end, as EMS/XMS can also be allocated before lowmem, if it's not aligned here.

stsp commented on src/arch/linux/mapping/mapfile.c:
```diff
@@ -267,6 +270,36 @@ static void *alloc_mapping_file(int cap, size_t mapsize)
+  if (!ptr)
+    return ptr;
+  /* smalloc_aligned aligns the size as well, so realloc in-place */
+  ptr = smrealloc(&pgmpool, ptr, mapsize);
```
Should be addressed in #1935

stsp commented:
OK, I think you just need to add the
"target" member to alloc ops and do
all the logic in mapping.c, instead of
adding another alloc ops.
smalloc-related hacks should be
replaced by #1935 
open_mapping_f() should not do what
you suggest in this patch. For example
open_mapping_shm() doesn't mmap
the pool at all. I think open_mapping_f()
should do the same. Smalloc can likely
be removed from there.

stsp commented:
I even think that realloc_mapping()
should also be moved to mapping.c,
instead of trying to do that in back-ends.
Doing that in back-ends results in
reallocating to always contiguous
regions of the underlying file. But all
we need is to mremap() with additional
anonymous pages, and then allocate
a few pages from shm with that "target"
addr. So adding "target" arg to alloc ops
will even allow to rework realloc_mapping()
properly.

bartoldeman commented:
> I even think that realloc_mapping() should also be moved to mapping.c, instead of trying to do that in back-ends. Doing that in back-ends results in reallocating to always contiguous regions of the underlying file. But all we need is to mremap() with additional anonymous pages, and then allocate a few pages from shm with that "target" addr. So adding "target" arg to alloc ops will even allow to rework realloc_mapping() properly.

That's possible but makes it more complex as well, as you still need to keep track then which (no longer contiguous) areas of the file are being mapped. So then a munmap() will need to also look up multiple file regions and mark those as free.

To save myself that headache I just memcpy to a new mapping if in-place resizing isn't possible. It's a rare case anyway.

The rest was all possible in #1936, using the pgalloc.c with a simple resize function added.

stsp commented:
> So then a munmap() will need to also look up multiple file regions and mark those as free.

Yep. And having a sparse allocator
(based on page bitmaps) will make
it too slow as you'll then need to map
and unmap page-by-page.

bartoldeman commented:
Rebased against devel, a lot simpler now with #1936 in.

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -671,6 +659,20 @@ void *alloc_mapping(int cap, size_t mapsize)
+{
+  Q__printf("MAPPING: alloc, cap=%s size=%#zx\n", cap, mapsize);
+  return alloc_mapping_common(cap, mapsize,
+	mmap(NULL, mapsize, PROT_NONE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0));
```
Wouldn't it be better to pass -1 as
"target", as we always do everywhere?

bartoldeman replied:
I don't think so, this is just a simple split of `alloc_mapping()` into `alloc_mapping()` and `alloc_mapping_common()`. If I pass `-1` to `alloc_mapping_common()` it'll have to check for it.

stsp replied:
Why would you check for it?
It will be passed to an alloc op and
no MAP_FIXED will be set.

bartoldeman replied:
oh I misunderstood, sure then the intermediate anon mapping is no longer necessary without huge pages.

bartoldeman replied:
Also renamed to `do_alloc_mapping()`, a bit more consistent with the other function names.

bartoldeman committed:
mapping: alloc_mapping_huge_page_aligned for SHM

This change introduces alloc_mapping_huge_page_aligned and removes
the now redundant alias_mapping_huge_page_aligned. It's used for
lowmem_base and vga.mem.base, and hence the intermediate mapping
for the VGAEMU LFB is no longer needed for KVM to align properly.

The guard page logic that was used for vga.mem.base in vgaemu.c
has been moved into mmap_mapping_huge_page_aligned(), so that there's
always a PROT_NOTE page in front of a huge page aligned mmap.

See also #1922

stsp commented:
Thanks!
